hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/ECUnderReplicationHandler.java

```diff
@@ -0,0 +1,244 @@
+        .getStorageSize(ScmConfigKeys.OZONE_SCM_CONTAINER_SIZE,
+            ScmConfigKeys.OZONE_SCM_CONTAINER_SIZE_DEFAULT, StorageUnit.BYTES);
+    this.nodeManager = nodeManager;
+  }
+
+  /**
+   * Identify a new set of datanode(s) to reconstruct the container and form the
+   * SCM command to send it to DN. In the case of decommission, it will just
+   * generate the replicate commands instead of reconstruction commands.
+   *
+   * @param replicas - An instance of ContainerReplicaCount, containing the
+   *                   current replica count and inflight adds and deletes
+   * @param pendingOps - Inflight replications and deletion metrics.
+   * @param result - Health check result.
+   * @param remainingMaintenanceRedundancy - represents that how many nodes go
+   *                                      into maintenance.
+   * @return Returns the key value pair of destination dn where the command gets
+   * executed and the command itself.
+   */
+  @Override
+  public Map<DatanodeDetails, SCMCommand> processAndCreateCommands(
+      final Set<ContainerReplica> replicas,
+      final ContainerReplicaPendingOps pendingOps,
+      final ContainerHealthResult result,
+      final int remainingMaintenanceRedundancy) {
+    ContainerInfo container = result.getContainerInfo();
+    ECReplicationConfig repConfig =
+        (ECReplicationConfig) container.getReplicationConfig();
+    List<ContainerReplicaOp> pendingOpsForContainer =
+        pendingOps.getPendingOps(container.containerID());
+    final ECContainerReplicaCount replicaCount =
+        new ECContainerReplicaCount(container, replicas, pendingOpsForContainer,
+            remainingMaintenanceRedundancy);
+
+    ContainerHealthResult.UnderReplicatedHealthResult containerHealthResult =
+        (ContainerHealthResult.UnderReplicatedHealthResult)
+            ecContainerHealthCheck
+            .checkHealth(container, replicas, pendingOpsForContainer,
+                remainingMaintenanceRedundancy);
+
+    LOG.debug("Handling under-replicated EC container: {}", container);
+    if (containerHealthResult
+        .isSufficientlyReplicatedAfterPending() || containerHealthResult
+        .getHealthState() != ContainerHealthResult.HealthState
+        .UNDER_REPLICATED) {
+      LOG.info("The container {} with replicas {} is sufficiently replicated",
+          container.getContainerID(), replicaCount.getReplicas());
+      // Assert replicaCount also telling it's sufficiently replicated.
+      return null;
+    }
+    final ContainerID id = container.containerID();
+    try {
+     // State is UNDER_REPLICATED
+      final List<DatanodeDetails> deletionInFlight = new ArrayList<>();
+      for (ContainerReplicaOp op : pendingOpsForContainer) {
+        if (op.getOpType() == ContainerReplicaOp.PendingOpType.DELETE) {
+          deletionInFlight.add(op.getTarget());
+        }
+      }
+      List<Integer> missingIndexes = replicaCount.unavailableIndexes(true);
```

Comment:
We need to subtract and inflight add replicas here.

Reply:
Sorry, cancel that - the "true" passed in here considers the pending add.

Reply:
When we pass flag true, isn't it subtracted already?